walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,163 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  /**
+   * Gets the available time slot for new meeting given existing meeting events.
+   * @param events Existing meeting events, assuming all events end before EOD.
+   * @param request Request to set up a new meeting
+   * @return If one or more time slots exists so that both mandatory and optional attendees can attend,
+   *         returns those time slots. Otherwise, returns the time slots that fit just the mandatory attendees.
+   *         Returns empty if duration is greater than a day.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+    if (!isValidParams(events, duration)) {
+      return Collections.unmodifiableList(new ArrayList<>());
+    }
+
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> attendeesWithOptional = new HashSet<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    // No attendees at all.
+    if (attendeesWithOptional.isEmpty()) {
+      return Collections.unmodifiableList(Arrays.asList(TimeRange.WHOLE_DAY));
+    }
+
+    List<TimeRange> occupiedTimeRange = new ArrayList<>();
+    List<TimeRange> occupiedTimeRangeWithOptional = new ArrayList<>();
+
+    for (Event event: events) {
+      getOccupiedTimeRange(attendees, occupiedTimeRange, event);
+
+      getOccupiedTimeRange(attendeesWithOptional, occupiedTimeRangeWithOptional, event);
+    }
+
+    List<TimeRange> availableMeetings = getAvailableTimeRange(occupiedTimeRangeWithOptional, duration);
+
+    // Has time slots that satisfy both mandatory and optional attendees.
+    if (!availableMeetings.isEmpty()) {
+      return Collections.unmodifiableList(availableMeetings);
```

Comment:
Maybe make getAvailableTimeRange returning immutableList and return it directly here?

Reply:
as stated above

Reply:
Fixed